fvm/transactionInvocator.go

```diff
@@ -152,54 +152,87 @@ func (i *TransactionInvocator) Process(
 	// 	panic(err)
 	// }
 
+	// try to deduct fees even if there is an error.
+	feesError := i.deductTransactionFees(env, proc)
+	if feesError != nil {
+		txError = feesError
+	}
+
 	// applying contract changes
 	// this writes back the contract contents to accounts
 	// if any error occurs we fail the tx
+	// this needs to happen before checking limits, so that contract changes are committed to the state
 	updatedKeys, err := env.Commit()
 	if err != nil && txError == nil {
 		txError = fmt.Errorf("transaction invocation failed: %w", err)
 	}
 
+	// if there is still no error check if all account storage limits are ok
 	if txError == nil {
 		txError = NewTransactionStorageLimiter().CheckLimits(env, sth.State().UpdatedAddresses())
 	}
 
-	if txError == nil {
-		txError = i.deductTransactionFees(env, proc)
-	}
-
-	proc.Logs = append(proc.Logs, env.Logs()...)
-	proc.ComputationUsed = proc.ComputationUsed + env.GetComputationUsed()
-
+	// it there was any transaction error clear changes and try to deduct fees again
 	if txError != nil {
-		// drop delta
+		// drop delta since transaction failed
 		childState.View().DropDelta()
 		// if tx fails just do clean up
 		programs.Cleanup(nil)
+		// log transaction as failed
 		i.logger.Info().
 			Str("txHash", proc.ID.String()).
 			Uint64("blockHeight", blockHeight).
 			Uint64("ledgerInteractionUsed", sth.State().InteractionUsed()).
 			Msg("transaction executed with error")
-		return txError
+
+		// reset env
+		env = NewTransactionEnvironment(*ctx, vm, sth, programs, proc.Transaction, proc.TxIndex, span)
+
+		// try to deduct fees again, to get the fee deduction events
```

Comment:
We could re-organise this so no second fees deduction code is run.
Something along the lines of
```delta = deductFees()
actualDelta = delta.NewChild()
err = runTx(actualDelta)
if err == nil {
  delta.Merge(actualDelta)
}
delta.Commit()
```